scrapy/downloadermiddlewares/robotstxt.py

```diff
@@ -57,7 +59,13 @@ def robot_parser(self, request, spider):
                 priority=self.DOWNLOAD_PRIORITY,
                 meta={'dont_obey_robotstxt': True}
             )
-            dfd = self.crawler.engine.download(robotsreq, spider)
+            # engine.download() can return an already-called deferred, e.g. if a
+            # middleware returns a response in process_request(). Using
+            # deferLater() ensures that the error callback isn't called
+            # immediately upon being added, so that it doesn't remove the key
+            # before we check for it.
+            dfd = deferLater(reactor, 0, self.crawler.engine.download,
```

Comment:
It solves the KeyError issue because it ensures dfd errbacks are not called while `robot_parser` function is executed. Error message can still be in a wrong place, but I don't see why would we want to make it further by using 0.1 instead of 0.